### PR TITLE
Support packed input for FA

### PR DESCRIPTION
**Pull request by ksivaman**
There are also some general refactoring of asserts and the backend choosing process to unify all of the attention mechanisms, among other smaller changes.

### CODE REVIEW

**Comment by ptrendx** [comment]
Let's remove the 0.10 label and move this to 0.11 - we are already past the deadline for 0.10.

**Comment by JimmyZhang12** [comment]
Why are the lines duplicated at https://github.com/NVIDIA/TransformerEngine/blob/fb0f2596183a65ddfc0fde319bb507eb6a8079bd/transformer_engine/pytorch/attention.py#L1169

and 

https://github.com/NVIDIA/TransformerEngine/blob/fb0f2596183a65ddfc0fde319bb507eb6a8079bd/transformer_engine/pytorch/attention.py#L1181
?

**Comment by ksivaman** [comment]
@cyanguwa @timmoon10 @ptrendx PTAL

**Comment by ksivaman** [comment]
Pipeline 8892190

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -380,32 +560,64 @@ def forward(
...
         ]
 
-        max_seqlen = seqlen
-        cu_seqlens = torch.arange(
-            0,
-            (batch_size + 1) * seqlen,
-            step=seqlen,
-            dtype=torch.int32,
-            device=query_layer.device)
+        if self.attn_mask_type != 'padding':
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:573]: The arbitrary mask case isn't handled. Perhaps we should add an assert that the attention mask type is either `"no_mask"` or `"causal"`.

File: transformer_engine/pytorch/softmax.py [hunk 2]
```diff
@@ -215,7 +215,7 @@ class FusedScaleMaskSoftmax(nn.Module):
     fused operation: scaling + mask + softmax
 
     Arguments:
-        attn_mask_type: attention mask type (pad or causal)
+        attn_mask_type: attention mask type ('padding', 'causal', or None)
```
Comment (timmoon10) [inline transformer_engine/pytorch/softmax.py:218]:
```suggestion
        attn_mask_type: attention mask type ('padding', 'causal', or 'no_mask')
```

File: tests/pytorch/test_onnx_export.py [hunk 3]
```diff
@@ -1038,14 +1038,14 @@ def test_export_layernorm_mlp(
...
-    (torch.float32,  False,    None),      # calls forward_torch_softmax
-    (torch.float32,  True,     None),      # calls forward_torch_softmax
-    (torch.float16,  False,    "causal"),  # calls ScaledUpperTriangMaskedSoftmax
-    (torch.float16,  True,     "padding"), # calls ScaledMaskedSoftmax
-    (torch.float16,  False,    "padding"), # calls ScaledSoftmax
-    (torch.bfloat16, False,    "causal"),  # calls ScaledUpperTriangMaskedSoftmax
-    (torch.bfloat16, True,     "padding"), # calls ScaledMaskedSoftmax
-    (torch.bfloat16, False,    "padding"), # calls ScaledSoftmax
+    (torch.float32,  False,    None),        # calls forward_torch_softmax
+    (torch.float32,  True,     None),        # calls forward_torch_softmax
```
Comment (timmoon10) [inline tests/pytorch/test_onnx_export.py:1042]:
```suggestion
    (torch.float32,  False,    "no_mask"),   # calls forward_torch_softmax
    (torch.float32,  True,     "no_mask"),   # calls forward_torch_softmax
```

File: transformer_engine/pytorch/attention.py [hunk 4]
```diff
@@ -714,8 +917,21 @@ class DotProductAttention(torch.nn.Module):
...
+                   TransformerEngine calculates and applies an upper triangular mask to
+                   the softmax input. An "`arbitrary`" mask is an arbitrary user defined mask
+                   broadcastable to the shape of softmax input. The "`padding`" mask is used
+                   for providing locations of padded tokens in the batch, which should be of
+                   the shape [batch_size, 1, 1, seq_len]. No mask is applied for the "`no_mask`"
+                   option. For the `"arbitrary"` and `"padding"` mask types, the argument
+                   :attr:`attention_mask` must be passed into `forward` call. The "`causal`"
+                   mask can also be applied in conjunction with "`padding`" mask by passing
+                   in multiple mask type as a comma separated string, for example,
+                   `attn_mask_type="causal,padding"`.
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:940]: Maybe in the future we can separate these mask types into two groups like PyTorch did: one for padding/arbitrary, which accepts a mask tensor input but the algorithm for padding or arbitrary would be different; and the other group is for is_causal, which doesn't need a mask tensor. We can just use two separate flags to represent them, on all levels of the implementation (DPA, Flash/FusedAttention, C level). 
https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html

File: transformer_engine/pytorch/attention.py [hunk 5]
```diff
@@ -52,6 +53,193 @@
...
+    Unpacks the attention mask type string and returns a single mask type
+    and a boolean for whether to apply causal mask. Also ensures that the
+    combination of masks passed in is supported by one of the attention
+    backends available.
+    """
+    mask_types = attn_mask_type.split(',')
+    assert (
+        all(mask_type in AttnMaskTypes for mask_type in mask_types)
+    ), f"Mask type {attn_mask_type} is not supported."
+
```
[the remark this answers is not in the document]
Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:229]: This check is taken care of in `_unpack_attn_mask_type` @cyanguwa

File: transformer_engine/pytorch/attention.py [hunk 6]
```diff
@@ -380,32 +560,64 @@ def forward(
...
+                step=max_seqlen_q,
+                dtype=torch.int32,
+                device=query_layer.device)
+            cu_seqlens_kv = torch.arange(
+                0,
+                (batch_size + 1) * max_seqlen_kv,
+                step=max_seqlen_kv,
+                dtype=torch.int32,
+                device=key_layer.device)
+            query_layer_packed, key_layer_packed, value_layer_packed = (
```
[the remark this answers is not in the document]
Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:592]: But ragged means something different right? Here we are basically taking the padded input and packing the valid tokens in every sample back to back for which `packed` makes sense to me.

File: transformer_engine/pytorch/attention.py (hunk 1 above)
Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:579]: That's correct it doesn't. But the check for masks is handled in `DotProductAttention` already, so by the time it reaches any specific backend, only supported masking types are present.

File: transformer_engine/pytorch/attention.py [hunk 7]
```diff
@@ -52,6 +53,193 @@
...
+    """
+    mask = mask.squeeze(1).squeeze(1)
+    reduced_mask = mask.sum(dim=1)
+    cu_seqlens = reduced_mask.cumsum(dim=0).to(torch.int32)
+    zero = torch.zeros(1, dtype=torch.int32, device="cuda")
+    cu_seqlens = torch.cat((zero, cu_seqlens))
+
+    mask = mask.reshape(-1)
+    indices = mask.nonzero()
+    indices = indices.unsqueeze(-1)
```
[the remark this answers is not in the document]
Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:74]: The shape of indices here is 3D, the `unsqueeze` changes the shape from `[n, 1]` to `[n, 1, 1]`.

[the remark this answers is not in the document]
Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:74]: The `reshape(-1)` makes it a 1D tensor and the nonzero creates a column tensor of shape `[n, 1]` with indices of all non zero locations.

File: tests/pytorch/test_onnx_export.py (hunk 3 above)
Comment (ksivaman) [reply tests/pytorch/test_onnx_export.py]: Done.

File: transformer_engine/pytorch/softmax.py (hunk 2 above)
Comment (ksivaman) [reply transformer_engine/pytorch/softmax.py]: Done.

File: transformer_engine/pytorch/attention.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py:592]: I initially thought `ragged` made sense because you were only packing the `valid` tokens back to back so that's what we were referring to as `ragged` on the kernels side. But I understand your thinking. I just think at some point we need to clean up the meanings of `packed` and `interleaved` a little bit. They are a little overloaded at the moment.

File: transformer_engine/pytorch/attention.py [hunk 8]
```diff
@@ -570,14 +575,15 @@ def forward(
             for x in [query_layer, key_layer, value_layer]
         ]
 
+        global _cu_seqlens_q, _cu_seqlens_kv, _indices_q, _indices_kv
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:578]: Why do we need these to be global variables? They don't seem to be used outside this function.

Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:578]: They're only calculated for `layer_number==1` if layer number is provided by the user and then cached for the other layers (as an optimization). Since it's just the cumulative sequence lengths the extra memory consumption is not an issue. @timmoon10

Comment (timmoon10) [reply transformer_engine/pytorch/attention.py:578]: This seems delicate, but I suppose `layer_number` should be considered an advanced optimization.

**Review by timmoon10 — APPROVED** [review]
LGTM

**Review by timmoon10 — APPROVED** [review]
Mostly LGTM

**Review by cyanguwa — APPROVED** [review]
Looks good to me

**Comment by cyanguwa** [comment]
Looks good to me. Two comments maybe - can you please address Tim's comment, and also keep the spacing style in lines such as 2046? Thanks!

**Comment by cyanguwa** [comment]
New fixes look good to me. Can merge once the CI passes. Thanks!
